### ISSUE TITLE
Erroneous recommendation for `types` exports

### ISSUE THREAD

**Issue by Conaclos** [bluwy/publint#46]
Hi!

Linting [my project with publint](https://publint.dev/@bare-ts/tools@0.13.0) raises a warning: the linter recommend adding a `types` exports.

However, if I follow this recommendation, this creates an issue according to a [linter of TypeScript team member](https://github.com/arethetypeswrong/arethetypeswrong.github.io): "Types are ESM, but implementation is CJS.". The only way to correctly export types when dual publishing cjs and esm is to remove the `types` exports and adding `.d.cts` / `.d.ts` / `.d.mts` along with corresponding files.

I understand the reasoning behind requiring `types` exports.
`publint` could detect dual publishing and then, in this particular case, recommend removing `types` exports.

**Comment by Conaclos** [comment]
Another solution could be requiring two `types` exports:

```
  "exports": {
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      },
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
  },

```

**Comment by bluwy** [comment]
Hmm, yeah I think I missed checking for the adjacent `.d.ts`, `.d.cts`, `.d.mts` when checking the lint rule.

Doing the opposite is an interesting idea. I don't particularly like that TypeScript wants people to publish `.d.ts`, `.d.cts`, `.d.mts` files separately, as if publishing types couldn't be harder. But anyways if anyone wants to contribute that, I wouldn't oppose it 🥲

**Comment by nvie** [comment]
Yeah, this is something that many packages are doing wrong, and it gets worse because most packages just cargo cult off of what other (popular) packages are doing.

I think it would hugely help if what @Conaclos is proposing were the recommended approach.

Basically, if you have a `require` _and_ an `import`, then I believe **it's _always_ a bug** if you have a `types` condition at the same level, too.

Reasoning:
- `require` and `import` cannot use the same file extension (because, if they did then CJS would be masquerading as ESM, or vice versa)
- Because `require` and `import` have different file extensions, their type definitions _must_ also have different type files

The solution, which would be nice to recommend as a best practice:

```json
{
  "exports": {
    ".": {
      "require": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.js",
      },
      "import": {
        "types": "./dist/index.d.mts",
        "default": "./dist/index.mjs",
      },
    }
  }
}
```

**Comment by bluwy** [comment]
Yeah being able to detect those cases would be good. I had a [small debate with Andrew](https://twitter.com/bluwyoo/status/1673516740431740929) and it's largely the reason why I didn't really push this. But at this point, I'm willing to have some sort of fix since it's what TS has set in stone.

I'm still somewhat on the side that having `types`, `import`, and `require` on the same level isn't always a bug though. A library could decide to only provide types for ESM or CJS. In this case, we can add information instead that the types will only work in ESM or CJS.

In the original issue, we can update our message more if the library exports both ESM and CJS. And also omit the warning if the the file has adjacent dts file (Maybe still good to recommend an explicit `types` condition instead? Just fix the recommendation to insert in the right condition nested)

**Comment by Conaclos** [comment]
> Basically, if you have a require and an import, then I believe it's always a bug if you have a types condition at the same level, too. It would be nice to add a lint rule that forbids this.

Not necessarily: order is important. I am currently using the following approach:

```json
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      },
      "types": "./dist/index.d.ts",
      "module": "./dist/index.js",
      "default": "./dist/index.js"
    }
  },
```

**Comment by nvie** [comment]
> I'm still somewhat on the side that having `types`, `import`, and `require` on the same level isn't always a bug though. A library could decide to only provide types for ESM or CJS. In this case, we can add information instead that the types will only work in ESM or CJS.

That's true. I agree, and providing a hint that the types won't work in one or the other would be nice to make it clear (because perhaps that's not the author's intention).

> In the original issue, we can update our message more if the library exports both ESM and CJS. And also omit the warning if the the file has adjacent dts file (Maybe still good to recommend an explicit `types` condition instead? Just fix the recommendation to insert in the right condition nested)

Agreed!

> Not necessarily: order is important. I am currently using the following approach:

A-ha, I see @Conaclos! Are you not getting the "types must come first" error? I was operating under the assumption that "types" would have to be defined before "require" according to the existing publint rule. If this setup doesn't lead to a linting error, then I agree this is also an option and it's indeed not always a bug.

**Comment by nvie** [comment]
Btw, @Conaclos – I just [contributed a rule last weekend](https://github.com/bluwy/publint/pull/49) that will fail on your `"module"` line. In this setup, I would recommend changing to:

```js
{
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "require": {
        "types": "./dist/index.d.cts",
        "module": "./dist/index.js",  // 👈 Move it to here, maybe?
        "default": "./dist/index.cjs"
      },
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
}
```

**Comment by Hebilicious** [comment]
Fix that is possibly related landed here https://github.com/bluwy/publint/issues/47

**Comment by bluwy** [comment]
Yeah, the "types must come first" with `import` and `require` conditions is fixed at #47. I think this issue is more specific to the recommendation detection, where:

1. It has `pkg.types`
2. But doesn't have `pkg.exports.types`

Since `"moduleResolution": "bundler"` would ignore `pkg.types` altogether. So we recommend some hints so `bundler` works, in this case the hints aren't perfect for dual publishing.

**Comment by Conaclos** [comment]
> Btw, @Conaclos – I just https://github.com/bluwy/publint/pull/49 that will fail on your "module" line. In this setup, I would recommend changing to:

`module` is bundler-specific condition (supported by webpack, esbuild, and others). It should not be in `require`.
By the way, I am not sure if `module` must be placed before `require`.
`module` makes sense only fore ESM? In this case the bundlers skip `require`?

So complex...

**Comment by nvie** [comment]
> So hard to get right...

You can say that again! 🥵

> `module` is bundler-specific condition (supported by webpack, esbuild, and others). It should not be in `require`. By the way, I am not sure if `module` must be placed before `require`. `module` makes sense only for ESM? In this case the bundlers skip `require`?

Not sure if that's true! Maybe it is? Now I don't know anymore 🥴

My understanding is that bundlers activate the `module` condition **_as well as_** `import` and `require` (depending on which construct is used to include the referenced module). So if a bundler encounters a `require('xxx')` call, they will normally resolve to `index.cjs` (through the `require` condition that they activated). Similarly, if they encounter an `import "xxx"` statement, they will normally resolve to `index.mjs` (through the `import` condition that they activated).

But bundlers _also_ activate the `module` condition, so that you have the opportunity to influence which modules they resolve to.

So assume that you have this:

```js
{
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "require": {
        "types": "./dist/index.d.cts",
        "module": "./dist/index.js",
        "default": "./dist/index.cjs"
      },
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
}
```

And your bundler processes this source code:

```js
const x = require('xxx');
```

1. Bundler sets `require` condition and `module` condition
2. Scanning top-to-bottom finds first match at `pkg.exports.require.module` → resolves to `./dist/index.js` (an ESM module!)
3. This is fine in a bundler context only, because bundlers possess the special power to `require()` an ESM module (unlike Node)

Similarly, if your bundler processes this source code:

```js
import * as x from 'xxx';
```

1. Bundler sets `import` condition and `module` condition
2. Top-to-bottom scanning happens.
3. Skips over `pkg.exports.require`, because `require` condition isn't met
4. Skips over `pkg.exports.types`, because `types` condition isn't met
5. Finds `pkg.exports.default` → resolves to `./dist/index.js` (also an ESM module!)

---

Maybe I'm wrong, but this is how I thought it works.

**Comment by nvie** [comment]
Btw, I'm mostly basing my understanding of this from the ["How conditions work"](https://esbuild.github.io/api/#how-conditions-work) section in the esbuild documentation. These parts in particular:

![Screen Shot 2023-07-04 at 23 55 05@2x](https://github.com/bluwy/publint/assets/83844/c0e206ae-588b-4945-85dc-79da36793ade)

**Comment by bluwy** [comment]
I've fixed this in https://github.com/bluwy/publint/commit/28f1ec3317b177b4dbfa9af9fff2e61908eea2a7 and https://github.com/bluwy/publint/commit/325b59fa092df03136a980f19c98aa2259192a2b. I've walked back a bit on this:

> I'm still somewhat on the side that having `types`, `import`, and `require` on the same level isn't always a bug though.

`publint` will now warn (for the root export only), if `import` and `require` both exist. Better to avoid bias wherever possible 😬 

You can start testing this out in https://publint.dev now. The site is updated but the package isn't published yet. Since this might affect many projects using `publint --strict`, I'm holding off the release a bit until tonight to write a better release note.

Closing this as fixed.

**Comment by bluwy** [comment]
This is now published in v0.2.0
